terra_notebook_utils/progress.py

```diff
@@ -0,0 +1,126 @@
+import sys
+import time
+import threading
+import sched
+from math import floor, ceil
+from contextlib import AbstractContextManager
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+
+class RateLimited:
+    """
+    Decorator to rate limit method calls. Raises `exception` if method is called more quickly than
+    `rate` times per second, or does nothing if exception is None.
+    """
+    def __init__(self, rate: float, exception: Exception=None):
+        self.period = 1 / rate
+        self.exception = exception
+        self.reset()
+
+    def reset(self):
+        self._last_call = time.time() - 10 * self.period
+
+    def __call__(self, func):
+        def wrapper(*args, **kwargs):
+            now = time.time()
+            if now - self._last_call >= self.period:
+                self._last_call = now
+                return func(*args, **kwargs)
+            elif self.exception is not None:
+                raise self.exception("Too soon")
+
+        wrapper.reset = self.reset
+        return wrapper
+
+
+class ProgressBar(AbstractContextManager):
+    def __init__(self, number_of_steps, prefix="", units="", size: int=None):
+        self.number_of_steps = number_of_steps
+        self.steps_completed = 0
+        self.prefix = prefix
+        self.units = units
+        self.size = size
+        self._length = 40
```

Comment:
Seems like a magic number?  Maybe a comment?

Reply:
Not magic so much as a choice for a reasonable length. I'll add a comment.